presto-common/src/main/java/com/facebook/presto/common/Utils.java

```diff
@@ -49,10 +56,190 @@ public static void checkArgument(boolean expression)
         }
     }
 
-    public static void checkArgument(boolean expression, String errorMessage)
+    public static void checkArgument(boolean expression, String message, Object... args)
     {
         if (!expression) {
-            throw new IllegalArgumentException(errorMessage);
+            throw new IllegalArgumentException(format(message, args));
+        }
+    }
+
+    /**
+     * Returns a supplier which caches the instance retrieved during the first call to {@code get()}
+     * and returns that value on subsequent calls to {@code get()}.
+     */
+    public static <T> Supplier<T> memoizedSupplier(Supplier<T> delegate)
+    {
+        if (delegate instanceof MemoizingSupplier) {
+            return delegate;
+        }
+        return new MemoizingSupplier<>(delegate);
+    }
+
+    static class MemoizingSupplier<T>
+            implements Supplier<T>
+    {
+        volatile Supplier<T> delegate;
+        volatile boolean initialized;
+        // "value" does not need to be volatile; visibility piggy-backs
+        // on volatile read of "initialized".
+        @Nullable T value;
+
+        MemoizingSupplier(Supplier<T> delegate)
+        {
+            this.delegate = requireNonNull(delegate);
+        }
+
+        @Override
+        public T get()
+        {
+            // A 2-field variant of Double Checked Locking.
+            if (!initialized) {
+                synchronized (this) {
+                    if (!initialized) {
+                        T t = delegate.get();
+                        value = t;
+                        initialized = true;
+                        // Release the delegate to GC.
+                        delegate = null;
+                        return t;
+                    }
+                }
+            }
+            return value;
+        }
+
+        @Override
+        public String toString()
+        {
+            Supplier<T> delegate = this.delegate;
+            return "Suppliers.memoize("
+                    + (delegate == null ? "<supplier that returned " + value + ">" : delegate)
+                    + ")";
+        }
+    }
+
+    public static ToStringHelper toStringHelper(Object self)
```

Comment:
Let's attribute this to Guava ?

Reply:
The guava dependency is not included in `presto-common` as the `presto-common` module is a dependency of the `presto-spi` module which we avoid putting additional dependencies so that when creating connectors or other plugins there it is less likely for downstream users of presto to encounter issues or conflicts

Reply:
OK, that's reasonable. However if we're forking this in, we also MUST include the tests and as well, and be prepared to support this for the foreseeable future. I'm not sure it's worth the effort. I'd be inclined to just manually build the strings in the toString method like we've been doing since 1995. :-)

Reply:
@ScrapCodes, as @elharo mentioned, shading is definitely a last resort and we should avoid it as much as possible. It's not worth doing. Guava is a huge codebase and would add a lot of bloat just for a few classes. I think vendoring these small utilities are useful though

Reply:
Fair enough, the can we have something like presto-vendored-guava module and use that?

Reply:
I removed ToStringHelper